### PR TITLE
HTTP signature authentication handler

### PR DESCRIPTION
**Pull request by rytmis**
This is the first part for #36. Right now, it's mostly plumbing:

- [x] Authentication handler
- [x] Interface for key retrieval
- [ ] Store keys to local database for faster access
  - [ ] Optionally cache most frequently used keys in memory
- [x] Infrastructure for validating signatures
- [x] Produce an authenticated ClaimsPrincipal

The idea is that once key retrieval is working, a signed HTTP request will produce a principal with the authentication scheme set to HttpSignature and the authenticated user name set to the fedi id of said user.

### CODE REVIEW

**Comment by jenniferplusplus** [comment]
I wanted to check in and see if you need help with anything? You made a ton of progress, and I'd love to get it merged. Not to rush you, just checking in.

**Comment by rytmis** [comment]
I've been a bit short on free time recently, so I don't know if I'm stuck or not. 😅 

I was trying to make the user signature validation work, and ran into Mastodon sending 127.0.0.1 as the host part of the key ID, which resulted in a TLS error in the Sandcastle setup. I haven't had time to figure out why that is yet. I expect after that's sorted, it's not too long until this is mergeable (although still lacking key persistence, caching and so on).

File: Letterbook.Core/CoreOptions.cs
Comment (jenniferplusplus) [inline Letterbook.Core/CoreOptions.cs:12]: There's an extension method that does a very similar thing. Would it work to use that instead?

I wanted to keep Options classes as just data containers if we can.

**Review by jenniferplusplus — APPROVED** [review]
This is great, thank you so much for doing it! I left a few questions. There's nothing that would need to block merging, but I'll give you a little time to make adjustments before I do that

File: Letterbook.Core/Models/SigningKey.cs
Comment (rytmis) [reply Letterbook.Core/Models/SigningKey.cs:10]: Thanks, I'll try to make the change later today

File: Letterbook.Core/Adapters/IHostSigningKeyProvider.cs
Comment (rytmis) [reply Letterbook.Core/Adapters/IHostSigningKeyProvider.cs:14]: The class, certainly. The interface, maybe not? For testability I'd probably prefer having a simple, single-purpose interface for key retrieval over a direct DbContext dependency. 

I spent about 30 minutes coming up with this to have something to work with, so I'm not overly attached to any of it. A db-backed instance actor certainly makes more sense than a disk-based key, but so far I've been hesitant to make a lot of changes to the entity model. Let me know what would make the most sense to you. :)

File: Letterbook.Core/Models/Profile.cs
Comment (rytmis) [reply Letterbook.Core/Models/Profile.cs:19]: I was mucking about with the debug API, trying to get a freshly reinitialized Mastodon to notice my Letterbook instance, and processing the Follow request blew up in EF Core.

I believe the chain of events was roughly:
- AP Client requests a Person-actor from Mastodon
- Result is deserialized into ASType
- ASType is mapped to Profile via AutoMapper, and there's no corresponding property to map from, so it's left uninitialized

It's the last part that invokes the private ctor, I think--I didn't verify it, but AFAICT there's no way to initialize a class without invoking _some_ constructor. In any case, [this change](https://github.com/Letterbook/Letterbook/blob/281d5d823a3055a961e8748c6774f440feca8ed5/Letterbook.Adapter.ActivityPub.Test/MapperTests.cs#L139) to a test was enough to trigger the issue.

If there's some other way you'd like to see this handled, let me know. :)

Comment (jenniferplusplus) [reply Letterbook.Core/Models/Profile.cs:19]: Oh, I just looked a the ast mapper again for the first time in a while. AutoMapper has a `.ConstructUsing()` method that let's you specify a method to create the object in the first place. I used that for our original not very good AP library, but that's not used anymore, so no wonder you didn't see it.

This is fine for now. It's clearly working. We can revisit if and when we have an actual need.

File: Letterbook.Core/CoreOptions.cs
Comment (rytmis) [reply Letterbook.Core/CoreOptions.cs:12]: It would, yes!

File: Letterbook.Api/Controllers/ActivityPub/ActorController.cs
Comment (rytmis) [reply Letterbook.Api/Controllers/ActivityPub/ActorController.cs:31]: I had to do some refresher reading, but there's multiple ways to vary the policy depending on the specific requirements and your preferred design.
- An obvious but clumsy alternative: multiple policies
- An imperative authentication alternative: explicitly invoke authentication with the Signature scheme and act based on the result, dropping the AuthorizeAttribute
- An authz-policy-based alternative: Custom-attribute-driven policy metadata per endpoint
- Imperative authz-policy-based alternative

For producing different outputs, purely declarative authorization would probably become unnecessarily convoluted. That would possibly be better served with imperative authorization checks, where the handlers can even be given access to an object representing the resource being accessed, e.g. some AP object.

If you want, say, a controller where there's one or two exceptions that should negotiate responses instead of outright forbidding the request, you could slap an `[AllowAnonymous]` on the methods, do an explicit check there and let the default work for everything else.

Another thing also occurred to me: should we ever actually allow a signed request through, if the verification fails? Right now, that results in no principal and the authz policy will deny the requests _when the policy is evaluated_, but it would allow to invoke any non-authorizing endpoints and the invalid signature would be ignored.

File: Letterbook.Adapter.ActivityPub/Client.cs
Comment (rytmis) [reply Letterbook.Adapter.ActivityPub/Client.cs:135]: BTW, since I decided against using the instance actor for client fetches (for now), and opted for an overload of Fetch that allows specifying the signing key instead, there's also the option to roll this and related changes back.

File: Letterbook.Core/Models/Profile.cs
Comment (rytmis) [reply Letterbook.Core/Models/Profile.cs]: I changed it anyway to be more consistent--I just added a similar CreateEmpty factory method for SigningKey as well.

File: Letterbook.Api/Controllers/ActivityPub/ActorController.cs
Comment (jenniferplusplus) [reply Letterbook.Api/Controllers/ActivityPub/ActorController.cs:31]: Um, good question. My first instinct is to say yes, we should allow (limited) anonymous access when verification fails. But, I think that puts a lot of faith in peer servers to actually refresh their caches and (and thus try again later, hopefully with fixed signatures). And I'm not sure that faith is warranted. So it might be better to fail loudly than to partially succeed.

Comment (rytmis) [reply Letterbook.Api/Controllers/ActivityPub/ActorController.cs:31]: I think I'll rework this a bit before merging, then:

- Put the signature validation in the middleware pipeline
- Always validate if signatures are present
- Fail requests containing invalid signatures with a 400
- Stick the successful result in a request feature or equivalent
- Have the handler query the result from there

This shouldn't be too difficult, but I think it will result in a pretty neat separation of concerns.

**Review by jenniferplusplus — APPROVED** [review]
I think you're allowed to merge, but let me know if not

**Pull request merged**
